discovery.md

```diff
@@ -0,0 +1,378 @@
+# CloudSubscriptions: Discovery - Version 0.1
+
+## Abstract
+
+CloudSubscriptions Discovery API is a vendor-neutral API specification for
+determining what events an Event Producer or middleware (Aggregator, Broker)
+has available, as well as how to subscribe to those events.
+
+## Status of this document
+
+This is a working draft. Breaking changes could be made in the next minor
+version.
+
+## Table of Contents
+
+- [Overview](#overview)
+- [Notations and Terminology](#notations-and-terminology)
+- [API Specification](#api-specification)
+- [Protocol Bindings](#protocol-bindings)
+- [Privacy & Security](#privacy-and-security)
+
+## Overview
+
+The goal of the CloudDiscovery API is to enable connection between consumers of
+events and producers / aggregators / brokers of events and to facilitate the
+creation of CloudSubscriptions. The discovery and subscription APIs are
+control-plane APIs to enable events to be delivered over the eventing
+data-plane.
+
+Discovery must allow for an event producer or middleware component to advertise
+the event types that are available, provide the necessary information to consume
+that event (schema / delivery protocol options), and the necessary information
+to create a subscription. The output of the API must be such that tooling can be
+built where all possible event producers and types aren’t known in advance.
+
+There are several discovery use cases to consider from the viewpoint of event
+consumers.
+
+1. What sources are available, and what event types do they produce?
+2. What event types are available, and from which providers.
+
+The second case becomes relevant if multiple providers support the same event
+types. Use case 1 is likely the dominant use case. Given the example of a public
+cloud provider where all producers produce events, there may be dozens of
+sources (producer systems) and hundreds of event types. The discovery funnel of
+producer first, then event types helps users navigate without having to see large
+lists of event types. Both of these cases require that filters can be used
+in the discovery API to narrow down the selection of available events.
+
+The CloudEvent source attribute is a potential source of high fanout. For
+example consider a blob storage system where each directory constitutes a
+distinct `source` attribute value. For this reason, source is not the starting
+point for discovery, but an attribute that can be used in creating subscriptions.
```

Comment:
It might be ok to say that `source` isn't the starting point, I would guess the producer is, so are you saying that people won't be able to traverse from "producer" to "source" to "events" via the discovery API?

Asking for events from just one bucket/dir seems like a fairly popular use case, I think it be natural to then ask for the list of events that a bucket/dir can generate? This way they don't see events that don't apply to other resources.

Reply:
Are you envisioning "source" being a "resource type" or a specific instance of a "resource type"? I'm guessing "resource type", which would reduce the fan-out, no?

Reply:
I think the most likely discovery flow is 

producer -> type -> source

Reply:
removing source from current RC of discovery